findtrieentries.go

```diff
@@ -0,0 +1,299 @@
+package veracity
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"fmt"
+	"strings"
+
+	"github.com/datatrails/go-datatrails-common/logger"
+	"github.com/datatrails/go-datatrails-merklelog/massifs"
+	"github.com/datatrails/go-datatrails-merklelog/mmr"
+	"github.com/google/uuid"
+	"github.com/urfave/cli/v2"
+)
+
+/**
+ * find trie Entries finds the trie entry associated with a given trie key
+ */
+
+const (
+	logTenantFlagName = "log-tenant"
+
+	logIDFlagName = "log-id"
+
+	domainFlagName = "domain"
+
+	appIDFlagName = "app-id"
+
+	asMMRIndexesFlagName = "as-mmrindexes"
+)
+
+var (
+	ErrNoLogTenant = fmt.Errorf("error, cannot find log tenant, please provide either %v or %v", logIDFlagName, logTenantFlagName)
+)
+
+// logIDToLogTenant converts the log id to the log tenant
+func logIDToLogTenant(logID string) (string, error) {
+
+	// first get the byte representation of the hex
+	logIdBytes, err := hex.DecodeString(logID)
+	if err != nil {
+		return "", err
+	}
+
+	// we don't know if its a log version 0 log id or a log version 1 log id
+
+	// attempt log version 1 first
+
+	// log version 1 is the byte representation of the uuid part of the log tenant
+	logIdUUid, err := uuid.ParseBytes(logIdBytes)
+	if err != nil {
+
+		// assume its log version 0 if it can't be parsed as bytes
+		// which is just utf-8 bytes of the log tenant string
+		return string(logIdBytes), nil
+	}
+
+	// if we get here we know its log version 1, so make the log tenant from the uuid
+	logTenant := fmt.Sprintf("tenant/%s", logIdUUid.String())
+
+	return logTenant, nil
+}
+
+// findTrieKeys searchs the log of the given log tenant for matches to the given triekeys
+// and returns the trie indexes (leaf indexes) of all the matches as well as the number of trie entries considered
+func findTrieKeys(log logger.Logger, massifReader MassifReader, logTenant string, trieKeys ...[]byte) ([]uint64, uint64, error) {
+
+	// get the head massif
+	headMassifContext, err := massifReader.GetHeadMassif(context.Background(), logTenant)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// find the number of massifs
+	massifCount := headMassifContext.Start.MassifIndex + 1
+
+	trieIndex := uint64(0)
+
+	matchingTrieIndexes := []uint64{}
+
```

Comment:
I would be inclined to allow the user to specify a massif range. Because typically I'd expect the user to have a good idea of reasonable range and also the full scan is going to get pretty expensive with some of the customers we are talking to now. It also means the memory for the function is bounded. To avoid having to "find the last massif" you could use end=-1 to indicate "all"

Reply:
great idea! will add a massif range

Reply:
added `massif-start` and `massif-end` please see PR description for different example uses :)

Reply:
is making --massif-end always optional and default to -1 problematic ?

Reply:
no, i suppose massif-start could also default to 0 :)

Reply:
done!